src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,889 @@
+        if (responseJson.has_field(L"error"))
+        {
+            throw "Time out.";
+        }
+        return responseJson;
+    }
+
+    // Method description:
+    // - helper function to acquire the user's Azure tenants
+    // Return value:
+    // - the response which contains a list of the user's Azure tenants
+    json::value AzureConnection::_GetTenants()
+    {
+        // Initialize the client
+        http_client tenantClient(_resourceUri);
+
+        // Initialize the request
+        http_request tenantRequest(L"GET");
+        tenantRequest.set_request_uri(L"tenants?api-version=2018-01-01");
+        _HeaderHelper(tenantRequest);
+
+        // Send the request and return the response as a json value
+        return _RequestHelper(tenantClient, tenantRequest);
+    }
+
+    // Method description:
+    // - helper function to refresh the access/refresh tokens
+    // Return value:
+    // - the response with the new tokens
+    json::value AzureConnection::_RefreshTokens()
+    {
+        // Initialize the client
+        http_client refreshClient(_loginUri);
+
+        // Initialize the request
+        http_request refreshRequest(L"POST");
+        refreshRequest.set_request_uri(_tenantID + L"/oauth2/token");
+        const auto body = L"client_id=" + AzureClientID + L"&resource=" + _wantedResource + L"&grant_type=refresh_token" + L"&refresh_token=" + _refreshToken;
+        refreshRequest.set_body(body, L"application/x-www-form-urlencoded");
+        refreshRequest.headers().add(L"User-Agent", resource);
+
+        // Send the request and return the response as a json value
+        return _RequestHelper(refreshClient, refreshRequest);
+    }
+
+    // Method description:
+    // - helper function to get the user's cloud shell settings
+    // Return value:
+    // - the user's cloud shell settings
+    json::value AzureConnection::_GetCloudShellUserSettings()
+    {
+        // Initialize client
+        http_client settingsClient(_resourceUri);
+
+        // Initialize request
+        http_request settingsRequest(L"GET");
+        settingsRequest.set_request_uri(L"providers/Microsoft.Portal/userSettings/cloudconsole?api-version=2018-10-01");
+        _HeaderHelper(settingsRequest);
+
+        return _RequestHelper(settingsClient, settingsRequest);
+    }
+
+    // Method description:
+    // - helper function to request for a cloud shell
+    // Return value:
+    // - the uri for the cloud shell
+    utility::string_t AzureConnection::_GetCloudShell()
+    {
+        // Initialize client
+        http_client cloudShellClient(_resourceUri);
+
+        // Initialize request
+        http_request shellRequest(L"PUT");
+        shellRequest.set_request_uri(L"providers/Microsoft.Portal/consoles/default?api-version=2018-10-01");
+        _HeaderHelper(shellRequest);
+        const auto innerBody = json::value::parse(U("{ \"osType\" : \"linux\" }"));
+        json::value body;
+        body[U("properties")] = innerBody;

```

Comment:
curious why you parsed json to get the inner value, then manually created the outer value. you could manually create both, or parse json (`{"properties":{"osType":"linux"}}`) for both :smile: